src/util/expr.cpp

```diff
@@ -539,3 +539,184 @@ const_unique_depth_iteratort exprt::unique_depth_cbegin() const
 { return const_unique_depth_iteratort(*this); }
 const_unique_depth_iteratort exprt::unique_depth_cend() const
 { return const_unique_depth_iteratort(); }
+
+/// We use the precendences that most readers expect
+/// (i.e., the ones you learn in primary school),
+/// and stay clear of the surprising ones that C has.
+static bool precedence(const exprt &a, const exprt &b)
+{
+  if((a.id()==ID_mult || a.id()==ID_div) &&
+     (b.id()==ID_plus || b.id()==ID_minus))
+    return true;
+
+  if((a.id()==ID_equal || a.id()==ID_notequal ||
+      a.id()==ID_lt || a.id()==ID_gt ||
+      a.id()==ID_le || a.id()==ID_ge) &&
+     (b.id()==ID_and || b.id()==ID_or))
+    return true;
+
+  return false;
+}
+
+/// This turns a multi-ary expression into a string,
+/// adding parentheses where indicated by the precedence.
+static std::string as_string_multi_ary(const exprt &src)
+{
+  bool first=true;
+
+  std::string result;
+
+  for(const auto &op : src.operands())
+  {
+    if(first)
+      first=false;
+    else
+    {
+      result+=' ';
+      result+=id2string(src.id());
+      result+=' ';
+    }
+
+    const bool need_parentheses=
+      !op.operands().empty() &&
+      !precedence(op, src);
+
+    if(need_parentheses)
+      result+='(';
+
+    result+=op.as_string();
+
+    if(need_parentheses)
+      result+=')';
+  }
+
+  return result;
+}
+
+/// This turns a binary expression into a string,
+/// adding parentheses where indicated by the precedence.
+static std::string as_string_binary(const exprt &src)
```

Comment:
`std::string as_string(const binary_exprt &src)` -- see above and below for discussion.

Reply:
done